mm2src/mm2_main/src/rpc.rs

```diff
@@ -316,60 +320,135 @@ pub extern "C" fn spawn_rpc(ctx_h: u32) {
 
     let ctx = MmArc::from_ffi_handle(ctx_h).expect("No context");
 
-    let rpc_ip_port = ctx.rpc_ip_port().unwrap();
+    let rpc_ip_port = ctx
+        .rpc_ip_port()
+        .unwrap_or_else(|err| panic!("Invalid RPC port: {}", err));
+    let is_https = ctx.https();
     // By entering the context, we tie `tokio::spawn` to this executor.
     let _runtime_guard = CORE.0.enter();
 
-    let server = Server::try_bind(&rpc_ip_port).unwrap_or_else(|_| panic!("Can't bind on {}", rpc_ip_port));
-    let make_svc = make_service_fn(move |socket: &AddrStream| {
-        let remote_addr = socket.remote_addr();
-        async move {
-            Ok::<_, Infallible>(service_fn(move |req: Request<Body>| async move {
-                let res = rpc_service(req, ctx_h, remote_addr).await;
-                Ok::<_, Infallible>(res)
-            }))
-        }
-    });
+    if is_https {
+        // Generate self-signed certificate
+        let subject_alt_names = vec!["localhost".to_string(), "127.0.0.1".to_string()];
+        let cert = generate_simple_self_signed(subject_alt_names)
+            .unwrap_or_else(|err| panic!("Can't generate self-signed for localhost and 127.0.0.1: {}", err));
+        let cert_der = cert
+            .serialize_der()
+            .unwrap_or_else(|err| panic!("Unable to serialize self-signed certificate: {}", err));
+
+        // Build TLS configuration
+        let privkey = PrivateKey(cert.serialize_private_key_der());
+        let cert = Certificate(cert_der);
+        let cert_chain = vec![cert];
+
+        // Create a TcpListener
+        let incoming =
+            AddrIncoming::bind(&rpc_ip_port).unwrap_or_else(|err| panic!("Can't bind on {}: {}", rpc_ip_port, err));
+        let acceptor = TlsAcceptor::builder()
+            .with_single_cert(cert_chain, privkey)
+            .unwrap_or_else(|err| panic!("Can't set certificate for TlsAcceptor: {}", err))
+            .with_all_versions_alpn()
+            .with_incoming(incoming);
+
+        let server = Server::builder(acceptor);
+        let make_svc = make_service_fn(move |socket: &TlsStream| {
+            let remote_addr = socket.remote_addr();
+            async move {
+                Ok::<_, Infallible>(service_fn(move |req: Request<Body>| async move {
+                    let res = rpc_service(req, ctx_h, remote_addr).await;
+                    Ok::<_, Infallible>(res)
+                }))
+            }
+        });
```

Comment:
Can we move this closure above of the `if-else` condition with taking `remote_addr` as an arg? I think we don't need to define this closure twice with different socket types.

Reply:
I am afraid passing `remote_addr` directly as a parameter to the closure in the `make_service_fn` function is not feasible due to the constraints imposed by the `serve` function implementation in the server builder. The `make_service_fn` function's generic type parameter `Target` is specifically defined in the `serve` function as the `Conn` type of the `Accept` trait implemented for both `TlsStream` and `AddrStream`.
I actually tried this before and some of the other suggestions below and they were hard to implement due to type constraints, but thinking more about this, I can do the following:
```rust
    let make_svc_fut = move |remote_addr: SocketAddr| async move {
        Ok::<_, Infallible>(service_fn(move |req: Request<Body>| async move {
            let res = rpc_service(req, ctx_h, remote_addr).await;
            Ok::<_, Infallible>(res)
        }))
    };

    macro_rules! make_svc {
        ($socket_type:ty) => {
            make_service_fn(move |socket: &$socket_type| {
                let remote_addr = socket.remote_addr();
                make_svc_fut(remote_addr)
            })
        };
    }
```
Where we pass the type that implements `Accept` as shown below
```rust
let server = server
            .http1_half_close(false)
            .serve(make_svc!(TlsStream))
            .with_graceful_shutdown(shutdown_fut);
```

Reply:
Macro solution seems great. It does what we need